src/libraries/Common/src/System/IO/PathInternal.cs

```diff
@@ -50,30 +50,26 @@ internal static int GetCommonPathLength(string first, string second, bool ignore
         /// <summary>
         /// Gets the count of common characters from the left optionally ignoring case
         /// </summary>
-        internal static unsafe int EqualStartingCharacterCount(string? first, string? second, bool ignoreCase)
+        internal static int EqualStartingCharacterCount(string? first, string? second, bool ignoreCase)
         {
             if (string.IsNullOrEmpty(first) || string.IsNullOrEmpty(second)) return 0;
 
-            int commonChars = 0;
-
-            fixed (char* f = first)
-            fixed (char* s = second)
+            for (int i = 0; i < first.Length; i++)
```

Comment:
What does perf look like if you just make this:
```csharp
return ignoreCase ?
    first.CommonPrefixLength(second, StringComparer.OrdinalIgnoreCase) :
    first.CommonPrefixLength(second);
```
(or something like that)

Reply:
@stephentoub actually, I'm not sure I understand what overload you're referring here. CommonPrefixLength accepts `ReadOnlySpan<T>` and the comparer has to be `IEqualityComparer<char>` which StringComparer.* can't be casted to?

Reply:
> and the comparer has to be IEqualityComparer<char> which StringComparer.* can't be casted to?

I wasn't thinking when I wrote `StringComparer.OrdinalIgnoreCase`; you'd need a comparer for chars, e.g.
```csharp
internal sealed class CharIgnoreCaseComparer : IEqualityComparer<char>
{
    public static CharIgnoreCaseComparer Instance { get; } = new();
    public bool Equals(char x, char y) => char.ToUpperInvariant(x) == char.ToUpperInvariant(y);
    public int GetHashCode(char obj) => char.ToUpperInvariant(obj).GetHashCode();
}
```
That's probably going to be measurably worse for ignoreCase=true, since now every comparison is an interface call. ignoreCase=false should hopefully be good, though.

Reply:
> We might need to add an overload that is specific to `char` and takes `StringComparison`

 _Originally posted by @terrajobst in [#64271](https://github.com/dotnet/runtime/issues/64271#issuecomment-1097103048)_